### ISSUE TITLE
EventLoopScheduler out of order

### ISSUE THREAD

**Issue by mail278** [dotnet/reactive#455]
Hi,

I have an issue with **EventLoopScheduler**.
EDIT: In fact in ObserveOn
I have two Subjects which synchronously generates events (first, second, first, second, etc.). If I observe these events directly (without any scheduler) it behaves as expected. If I observe these subjects by EventLoopScheduler, **sometimes two events swap order**.
**The problem is strictly related to the state of EventLoopScheduler when it is "starving"** - when it just finished all work and tries to start new batch of events. For this I always create few events and then wait 1ms to be sure the queue is empty and repeat.

this is RoslynPad sample code

```
#r "$NuGet\xunit\1.8.0.1545\lib\xunit.dll"
#r "$NuGet\System.Reactive.Core\3.1.1\lib\net46\System.Reactive.Core.dll"
#r "$NuGet\System.Reactive.Interfaces\3.1.1\lib\net45\System.Reactive.Interfaces.dll"
#r "$NuGet\System.Reactive.Linq\3.1.1\lib\net46\System.Reactive.Linq.dll"
#r "$NuGet\System.Reactive.PlatformServices\3.1.1\lib\net46\System.Reactive.PlatformServices.dll"
#r "$NuGet\System.Reactive.Windows.Threading\3.1.1\lib\net45\System.Reactive.Windows.Threading.dll"
#r "System.Windows"
#r "WindowsBase"


using System.Reactive.Concurrency;
using System.Reactive.Linq;
using System.Reactive.Subjects;

var a = new Subject<int>();
var b = new Subject<int>();
var scheduler = new EventLoopScheduler();

int counter = 0;
int receiver = 0;

// SUBSCRIBE
a.ObserveOn(scheduler).Subscribe(i =>
{
    // just check if the events are in order
    if (i != ++receiver) ("E" + i + " != " + receiver).Dump();
}, () => "FINISHED A".Dump());

b.ObserveOn(scheduler).Subscribe(i =>
{
    // just check if the events are in order
    if (i != -(++receiver)) ("F" + i + " != " + -receiver).Dump();
}, () => "FINISHED B".Dump());


// PUBLISH
for (int ii = 0; ii < 10000; ii++)
{
    // fill the EventLoopMachine queue
    a.OnNext(++counter);
    b.OnNext(-(++counter));
    a.OnNext(++counter);
    b.OnNext(-(++counter));
    a.OnNext(++counter);
    b.OnNext(-(++counter));
    a.OnNext(++counter);
    b.OnNext(-(++counter));
    a.OnNext(++counter);
    b.OnNext(-(++counter));
    // let the queue time to finish
    System.Threading.Thread.Sleep(1);
}

a.OnCompleted();
b.OnCompleted();

counter.Dump();
receiver.Dump();

```

EDIT: First I thought that the issue is in EventLoopScheduler, but after further investigation the issue is found in ObserveOn method (or in usage when I observe two IObservables with one eventScheduler).

when I change the code this way it works ok

this is out of order
```
a.ObserveOn(scheduler).Subscribe(i =>
{
  DoCheck();
});
```

this is in order
```
a.Subscribe(i =>
{
  scheduler.Schedule(() =>
  {
    DoCheck();
  });
});
```

**Comment by akarnokd** [comment]
I can't reproduce this issue with **4.1.0-preview.41**. It prints:

```
100000
100000
FINISHED A
FINISHED B
Press any key to continue . . .
```

**Comment by mail278** [comment]
Hi, I can confirm it is fixed in 4.1.0. Thanks!
